# Battle Of Neighborhood - Bangalore

#### Import all the required Libraries.

In [119]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

from bs4 import BeautifulSoup
import requests
import csv

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

!conda install -c conda-forge geocoder --yes # install geocoder
import geocoder

print('Libraries imported.')

Libraries imported.


#### Get the Postal Code and Neighborhoods of Bangalore from the mapsofindia website using beatifulsoup

In [120]:
source = requests.get('https://www.mapsofindia.com/pincode/india/karnataka/bangalore/').text

bsoup = BeautifulSoup(source,'lxml')

tablesoup = bsoup.find('table', class_ = 'extrtable')
#print(tablesoup.prettify())

In [121]:
# Create a CSV File by iterating the required tags to get the Pincode & Neighborhood

bangalore_csv = open('bangalore_neighborhood.csv', 'w')
csv_writer = csv.writer(bangalore_csv)
csv_writer.writerow(['Neighborhood','Pincode']) 

for tr in tablesoup.find_all('tr'):
#    print(tr)   
#tr = tablesoup.find('tr')
#print(tr.find('a'))
    if tr.find('a') != None :
        Neighborhood = tr.find('a').text
        Pincode = tr.find('b').text.split()[0]
#        print(Neighborhood, Pincode)
        csv_writer.writerow([Neighborhood,Pincode]) 

bangalore_csv.close()

In [122]:
#Create the dataframe from the CSV
df_b = pd.read_csv('bangalore_neighborhood.csv')

df_b.head()

Neighborhood  Pincode
0  A F station yelahanka   560063
1                Adugodi   560030
2                  Agara   560034
3                  Agram   560007
4     Air Force hospital   560007

In [123]:
df_b.shape

(294, 2)

In [124]:
# Neighborhoods with same Pincode are combined with a comma

grouped = df_b.groupby(['Pincode'])['Neighborhood'].apply(', '.join)

groupedDF = grouped.to_frame().reset_index()

print(groupedDF.shape)
groupedDF.head()

(99, 2)


Pincode                                       Neighborhood
0   560001  Bangalore Bazaar, Bangalore., Brigade Road, Cu...
1   560002  Avenue Road, Bangalore City, Bangalore Corpora...
2   560003  Aranya Bhavan, Malleswaram, Palace Guttahalli,...
3   560004  Basavanagudi, Lalbagh West, Mavalli, Pasmpamah...
4   560005                          Fraser Town, Jeevanahalli

### This code was used to retrieve lat,lon of all the postal code for one time due to limitation of google gecoder usage and saved the data in a CSV for reuse.

import geocoder # import geocoder

bangalore_LLcsv = open('bangalore_LL.csv', 'w')
csv_writer = csv.writer(bangalore_LLcsv)
csv_writer.writerow(['Pincode','latitude','longitude']) 

for index, row in groupedDF.iterrows():
    # initialize your variable to None
    lat_lng_coords = None
    postalcode = groupedDF.loc[index,'Pincode']
    # loop until you get the coordinates
    while(lat_lng_coords is None):
      g = geocoder.google('{}, Bangalore, Karnataka'.format(postalcode))
      lat_lng_coords = g.latlng

    latitude = lat_lng_coords[0]
    longitude = lat_lng_coords[1]
    #print(postalcode, '-> ', latitude, longitude)
    csv_writer.writerow([postalcode,latitude,longitude])

bangalore_LLcsv.close()
print('ll Completed')

In [125]:
# Create a DataFrame for the Lat, Lon data from the saved CSV File
bdf_geoloc = pd.read_csv('Bangalore_LL.csv')
bdf_geoloc.columns=['Pincode','Latitude','Longitude']
bdf_geoloc.head()

Pincode   Latitude  Longitude
0   560001  12.976594  77.599271
1   560002  12.963521  77.582106
2   560003  13.001914  77.571336
3   560004  12.943751  77.573760
4   560005  12.997989  77.622651

In [126]:
# Merge the 2 dataframe comparing the Postalcode column
bdf_latlon = pd.merge(groupedDF, bdf_geoloc, on='Pincode')

bdf_latlon.head(10)

Pincode                                       Neighborhood   Latitude  \
0   560001  Bangalore Bazaar, Bangalore., Brigade Road, Cu...  12.976594   
1   560002  Avenue Road, Bangalore City, Bangalore Corpora...  12.963521   
2   560003  Aranya Bhavan, Malleswaram, Palace Guttahalli,...  13.001914   
3   560004  Basavanagudi, Lalbagh West, Mavalli, Pasmpamah...  12.943751   
4   560005                          Fraser Town, Jeevanahalli  12.997989   
5   560006                    J.C.nagar, Training Command iaf  13.013841   
6   560007                          Agram, Air Force hospital  12.957917   
7   560008       H.A.l ii stage, Hulsur Bazaar, Someswarapura  12.981994   
8   560009  Bangalore Dist offices bldg, K. g. road, Subha...  12.977029   
9   560010  Bhashyam Circle, Industrial Estate, Rajajinaga...  12.984981   

   Longitude  
0  77.599271  
1  77.582106  
2  77.571336  
3  77.573760  
4  77.622651  
5  77.587667  
6  77.630912  
7  77.625613  
8  77.577305  
9  77.553344

#### Use geopy library to get the latitude and longitude values of Bangalore.

In [127]:
address = 'Bangalore, India'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Bangalore are {}, {}.'.format(latitude, longitude))

/home/jupyterlab/conda/lib/python3.6/site-packages/geopy/geocoders/osm.py:143: UserWarning: Using Nominatim with the default "geopy/1.17.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  UserWarning


The geograpical coordinate of Bangalore are 12.9791198, 77.5912997.


### Show the Neighborhoods in Bangalore Map

In [128]:
# create map of Bangalore using latitude and longitude values

map_bangalore = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, neighborhood in zip(bdf_latlon['Latitude'], bdf_latlon['Longitude'], bdf_latlon['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_bangalore)  
    
map_bangalore

### Define Foursquare Credentials and Version

In [129]:
CLIENT_ID = 'KAK50CKL4R2TKOFNRTELIGQRHURR2EKB00DK2WV1UM3TEZBP' # your Foursquare ID
CLIENT_SECRET = 'JKSD5YXIQSPBNSYEVOQW0ZFIVT1OHMX12IJVWJWREED5SUPP' # your Foursquare Secret
VERSION = '20180930' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: KAK50CKL4R2TKOFNRTELIGQRHURR2EKB00DK2WV1UM3TEZBP
CLIENT_SECRET:JKSD5YXIQSPBNSYEVOQW0ZFIVT1OHMX12IJVWJWREED5SUPP


#### Let's explore the first neighborhood in our dataframe.

In [130]:
bdf_latlon.loc[0, 'Neighborhood']

'Bangalore Bazaar, Bangalore., Brigade Road, Cubban Road, Dr. ambedkar veedhi, Highcourt, Legislators Home, Mahatma Gandhi road, Narayan Pillai street, Rajbhavan, Vidhana Soudha'

In [131]:
neighborhood_latitude = bdf_latlon.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = bdf_latlon.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood = bdf_latlon.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Bangalore Bazaar, Bangalore., Brigade Road, Cubban Road, Dr. ambedkar veedhi, Highcourt, Legislators Home, Mahatma Gandhi road, Narayan Pillai street, Rajbhavan, Vidhana Soudha are 12.9765944, 77.5992708.


#### Now, let's get the top 100 venues that are in the 1st Neighborhood within a radius of 1000 meters.

In [132]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 1000 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=KAK50CKL4R2TKOFNRTELIGQRHURR2EKB00DK2WV1UM3TEZBP&client_secret=JKSD5YXIQSPBNSYEVOQW0ZFIVT1OHMX12IJVWJWREED5SUPP&v=20180930&ll=12.9765944,77.5992708&radius=1000&limit=100'

Send the GET request and examine the resutls

In [133]:
results = requests.get(url).json()
results
type(results)

dict

In [134]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [135]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name           categories        lat        lng
0    M. Chinnaswamy Stadium       Cricket Ground  12.978144  77.599223
1  Hard Rock Cafe Bengaluru  American Restaurant  12.976154  77.601616
2              Corner House       Ice Cream Shop  12.972979  77.599971
3               Cubbon Park                 Park  12.977042  77.595277
4                 Rice Bowl   Chinese Restaurant  12.974596  77.599470

In [136]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

100 venues were returned by Foursquare.


### Now creating a function to repeat the same process to all the Nieghborhoods in Bangalore

In [137]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Call the Function to parse each neighborhood and create the bangalore_venues DF

In [138]:
bangalore_venues = getNearbyVenues(names=bdf_latlon['Neighborhood'],
                                   latitudes=bdf_latlon['Latitude'],
                                   longitudes=bdf_latlon['Longitude']
                                  )

Bangalore Bazaar, Bangalore., Brigade Road, Cubban Road, Dr. ambedkar veedhi, Highcourt, Legislators Home, Mahatma Gandhi road, Narayan Pillai street, Rajbhavan, Vidhana Soudha
Avenue Road, Bangalore City, Bangalore Corporation building, Bangalore Fort, Basavaraja Market, Cahmrajendrapet, Narasimjharaja Road, New Tharaggupet, Sri Jayachamarajendra road
Aranya Bhavan, Malleswaram, Palace Guttahalli, Swimming Pool extn, Venkatarangapura, Vyalikaval Extn
Basavanagudi, Lalbagh West, Mavalli, Pasmpamahakavi Road, Shankarpura, Visveswarapuram
Fraser Town, Jeevanahalli
J.C.nagar, Training Command iaf
Agram, Air Force hospital
H.A.l ii stage, Hulsur Bazaar, Someswarapura
Bangalore Dist offices bldg, K. g. road, Subhashnagar
Bhashyam Circle, Industrial Estate, Rajajinagar, Rajajinagar I block, Rajajinagar Ivth block
Jayangar Iii block, Madhavan Park
Science Institute
Govindapalya, H M t, Jalahalli, Jalahalli Village
Jalahalli East
Jalahalli West, Kamagondanahalli
Doorvaninagar, Krishnarajapuram

KeyError: 'groups'

In [139]:
print(bangalore_venues.shape)
bangalore_venues.head()

(2366, 7)


Neighborhood  Neighborhood Latitude  \
0  Bangalore Bazaar, Bangalore., Brigade Road, Cu...              12.976594   
1  Bangalore Bazaar, Bangalore., Brigade Road, Cu...              12.976594   
2  Bangalore Bazaar, Bangalore., Brigade Road, Cu...              12.976594   
3  Bangalore Bazaar, Bangalore., Brigade Road, Cu...              12.976594   
4  Bangalore Bazaar, Bangalore., Brigade Road, Cu...              12.976594   

   Neighborhood Longitude                     Venue  Venue Latitude  \
0               77.599271    M. Chinnaswamy Stadium       12.978144   
1               77.599271  Hard Rock Cafe Bengaluru       12.976154   
2               77.599271              Corner House       12.972979   
3               77.599271               Cubbon Park       12.977042   
4               77.599271                 Rice Bowl       12.974596   

   Venue Longitude       Venue Category  
0        77.599223       Cricket Ground  
1        77.601616  American Restaurant  
2        77.599971       Ice Cream Shop  
3        77.595277                 Park  
4        77.599470   Chinese Restaurant

Let's check how many venues were returned for each neighborhood

In [140]:
bangalore_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
A F station yelahanka                                                   5   
Adugodi, Hosur Road                                                     9   
Agara, Kendriya Sadan, Koramangala, Koramangala...                     56   
Agram, Air Force hospital                                               4   
Amruthahalli, Kodigehalli, Sahakaranagar P.o                           27   
Anandnagar, H.A. farm, Hebbal Kempapura                                14   
Anekal, Anekalbazar, Bestamaranahalli, Harogadd...                      1   
Arabic College, Devarjeevanahalli, Nagavara, Ra...                      5   
Aranya Bhavan, Malleswaram, Palace Guttahalli, ...                     51   
Ashoknagar, Banashankari, Dasarahalli, State Ba...                     26   
Attur, Bsf Campus yelahanka, Crpf Campus yelaha...                     15   
Austin Town, Viveknagar                                                 5   
Avalahalli, Bapujinagar, Byatarayanapura, Deepa...                      7   
Avani Sringeri mutt, Basaveswaranagar Ii stage,...                     16   
Avenue Road, Bangalore City, Bangalore Corporat...                     18   
B Sk ii stage, Jayanagar West, Padmanabhnagar, ...                     15   
Bagalgunte, Nagasandra                                                  4   
Bagalur, Bandikodigehalli, Doddagubbi, Kannur                           1   
Balepete, Chickpet                                                     28   
Banashankari Iii stage, Girinagar, Hosakerehall...                     21   
Banaswadi, Horamavu, Jalavayuvihar, Kalyanagar                         42   
Bangalore Air port, Nal, Vimapura                                      26   
Bangalore Bazaar, Bangalore., Brigade Road, Cub...                    100   
Bangalore Dist offices bldg, K. g. road, Subhas...                     42   
Bangalore Sub fgn post, Cmp Centre and school, ...                    100   
Bannerghatta                                                            1   
Bannerghatta Road, Hulimavu, Mico Layout, Mount...                     32   
Basavanagudi, Lalbagh West, Mavalli, Pasmpamaha...                     53   
Basaveshwaranagar, K H b colony, Kamakshipalya                         22   
Bellandur                                                              26   
Benson Town                                                            20   
Bettahalsur, Chikkajala, Doddajala, Hunasamaran...                      4   
Bhashyam Circle, Industrial Estate, Rajajinagar...                     14   
Bhattarahalli, Bidrahalli, Medihalli, Mundur, T...                      6   
Bnagalore Viswavidalaya, Mallathahalli, Ullalu ...                      2   
Bommanahalli, Hongasandra, Madivala                                    11   
C.V.raman nagar                                                        24   
Chamrajpet, Chamrajpet Bazar, Vittalnagar                              12   
Chandra Lay out, Vijayanagar, Vijayanagar East                         28   
Chikkabettahalli, Vidyaranyapura                                       11   
Chikkalasandra, Subramanyapura                                          4   
Devanagundi, Kadugodi, Kalkunte, Kannamangala, ...                      4   
Devasandra, Krishnarajapuram                                            5   
Dharmaram College, Taverekere                                          79   
Doddakallasandra, Konanakunte                                           4   
Doddanekkundi, Kundalahalli, Marathahalli Colon...                     50   
Domlur                                                                 54   
Dommasandra, Handenahalli, Kugur, Sarjapura                             2   
Doorvaninagar, Krishnarajapuram R s, Ramamurthy...                      4   
Electronics City                                                       16   
Fraser Town, Jeevanahalli                       

### Let's find out how many unique categories can be curated from all the returned venues

In [141]:
print('There are {} uniques categories.'.format(len(bangalore_venues['Venue Category'].unique())))

There are 206 uniques categories.


### Let's find out the top 10 Venue Categories in the neighborhoods

In [142]:
bangalore_venues['Venue Category'].value_counts().head(10)

Indian Restaurant       428
Café                    161
Fast Food Restaurant     85
Ice Cream Shop           72
Pizza Place              70
Department Store         70
Coffee Shop              68
Hotel                    64
Bakery                   59
Chinese Restaurant       58
Name: Venue Category, dtype: int64

### Analyze Each Neighborhood

In [143]:
# one hot encoding
bangalore_onehot = pd.get_dummies(bangalore_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
bangalore_onehot['Neighborhood'] = bangalore_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [bangalore_onehot.columns[-1]] + list(bangalore_onehot.columns[:-1])
bangalore_onehot = bangalore_onehot[fixed_columns]

bangalore_onehot.head()

Neighborhood  Accessories Store  \
0  Bangalore Bazaar, Bangalore., Brigade Road, Cu...                  0   
1  Bangalore Bazaar, Bangalore., Brigade Road, Cu...                  0   
2  Bangalore Bazaar, Bangalore., Brigade Road, Cu...                  0   
3  Bangalore Bazaar, Bangalore., Brigade Road, Cu...                  0   
4  Bangalore Bazaar, Bangalore., Brigade Road, Cu...                  0   

   Afghan Restaurant  American Restaurant  Andhra Restaurant  Arcade  \
0                  0                    0                  0       0   
1                  0                    1                  0       0   
2                  0                    0                  0       0   
3                  0                    0                  0       0   
4                  0                    0                  0       0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0            0           0                    0                 0   
1            0           0                    0                 0   
2            0           0                    0                 0   
3            0           0                    0                 0   
4            0           0                    0                 0   

   Athletics & Sports  Auto Garage  Auto Workshop  BBQ Joint  Badminton Court  \
0                   0            0              0          0                0   
1                   0            0              0          0                0   
2                   0            0              0          0                0   
3                   0            0              0          0                0   
4                   0            0              0          0                0   

   Bagel Shop  Bakery  Bar  Basketball Court  Bed & Breakfast  \
0           0       0    0                 0                0   
1           0       0    0                 0                0   
2           0       0    0                 0                0   
3           0       0    0                 0                0   
4           0       0    0                 0                0   

   Bengali Restaurant  Bike Shop  Bistro  Boarding House  Boat or Ferry  \
0                   0          0       0               0              0   
1                   0          0       0               0              0   
2                   0          0       0               0              0   
3                   0          0       0               0              0   
4                   0          0       0               0              0   

   Bookstore  Botanical Garden  Boutique  Bowling Alley  Breakfast Spot  \
0          0                 0         0              0               0   
1          0                 0         0              0               0   
2          0                 0         0              0               0   
3          0                 0         0              0               0   
4          0                 0         0              0               0   

   Brewery  Bridge  Bubble Tea Shop  Burger Joint  Burrito Place  Bus Station  \
0        0       0                0             0              0            0   
1        0       0                0             0              0            0   
2        0       0                0             0              0            0   
3        0       0                0             0              0            0   
4        0       0                0             0              0            0   

   Bus Stop  Business Service  Butcher  Cafeteria  Café  Candy Store  \
0         0                 0        0          0     0            0   
1         0                 0        0          0     0            0   
2         0                 0        0          0     0            0   
3         0                 0        0          0     0            0   
4         0                 0        0          0     0            0   

   Cantonese Restaurant  Capitol Building

In [144]:
bangalore_onehot.shape

(2366, 207)

### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [145]:
bangalore_grouped = bangalore_onehot.groupby('Neighborhood').mean().reset_index()
bangalore_grouped.head()

Neighborhood  Accessories Store  \
0                              A F station yelahanka                0.0   
1                                Adugodi, Hosur Road                0.0   
2  Agara, Kendriya Sadan, Koramangala, Koramangal...                0.0   
3                          Agram, Air Force hospital                0.0   
4       Amruthahalli, Kodigehalli, Sahakaranagar P.o                0.0   

   Afghan Restaurant  American Restaurant  Andhra Restaurant  Arcade  \
0                0.0             0.000000           0.000000     0.0   
1                0.0             0.000000           0.000000     0.0   
2                0.0             0.017857           0.017857     0.0   
3                0.0             0.000000           0.000000     0.0   
4                0.0             0.000000           0.000000     0.0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0          0.0         0.0                  0.0               0.0   
1          0.0         0.0                  0.0               0.0   
2          0.0         0.0                  0.0               0.0   
3          0.0         0.0                  0.0               0.0   
4          0.0         0.0                  0.0               0.0   

   Athletics & Sports  Auto Garage  Auto Workshop  BBQ Joint  Badminton Court  \
0                 0.0          0.0            0.0   0.000000              0.0   
1                 0.0          0.0            0.0   0.000000              0.0   
2                 0.0          0.0            0.0   0.017857              0.0   
3                 0.0          0.0            0.0   0.000000              0.0   
4                 0.0          0.0            0.0   0.000000              0.0   

   Bagel Shop    Bakery       Bar  Basketball Court  Bed & Breakfast  \
0         0.0  0.000000  0.000000               0.0              0.0   
1         0.0  0.000000  0.000000               0.0              0.0   
2         0.0  0.053571  0.017857               0.0              0.0   
3         0.0  0.000000  0.000000               0.0              0.0   
4         0.0  0.000000  0.000000               0.0              0.0   

   Bengali Restaurant  Bike Shop  Bistro  Boarding House  Boat or Ferry  \
0                 0.0        0.0     0.0             0.0            0.0   
1                 0.0        0.0     0.0             0.0            0.0   
2                 0.0        0.0     0.0             0.0            0.0   
3                 0.0        0.0     0.0             0.0            0.0   
4                 0.0        0.0     0.0             0.0            0.0   

   Bookstore  Botanical Garden  Boutique  Bowling Alley  Breakfast Spot  \
0        0.0               0.0       0.0            0.0        0.000000   
1        0.0               0.0       0.0            0.0        0.000000   
2        0.0               0.0       0.0            0.0        0.017857   
3        0.0               0.0       0.0            0.0        0.000000   
4        0.0               0.0       0.0            0.0        0.000000   

    Brewery  Bridge  Bubble Tea Shop  Burger Joint  Burrito Place  \
0  0.000000     0.0         0.000000      0.000000            0.0   
1  0.000000     0.0         0.000000      0.000000            0.0   
2  0.000000     0.0         0.000000      0.000000            0.0   
3  0.000000     0.0         0.000000      0.000000            0.0   
4  0.037037     0.0         0.037037      0.037037            0.0   

   Bus Station  Bus Stop  Business Service  Butcher  Cafeteria      Café  \
0     0.000000       0.0          0.000000      0.0        0.0  0.000000   
1     0.111111       0.0          0.111111      0.0        0.0  0.333333   
2     0.000000       0.0          0.000000      0.0        0.0  0.160714   
3     0.000000       0.0          0.000000      0.0        0.0  0.000000   
4     0.111111       0.0          0.000000      0.0        0.0  0.037037   

   Candy Store  Cantonese Restaurant  Capitol B

In [146]:
bangalore_grouped.shape

(96, 207)

## Create the bangalore_data Dataframe Removing the Neighbors where there were no venues found

In [147]:
bangalore_data = pd.merge(bangalore_grouped[['Neighborhood']], bdf_latlon, on='Neighborhood')

print(bangalore_data.shape)
bangalore_data.head()

(96, 4)


Neighborhood  Pincode   Latitude  \
0                              A F station yelahanka   560063  13.129087   
1                                Adugodi, Hosur Road   560030  12.944150   
2  Agara, Kendriya Sadan, Koramangala, Koramangal...   560034  12.926138   
3                          Agram, Air Force hospital   560007  12.957917   
4       Amruthahalli, Kodigehalli, Sahakaranagar P.o   560092  13.064104   

   Longitude  
0  77.614226  
1  77.607623  
2  77.622109  
3  77.630912  
4  77.593121

### Let's print each neighborhood along with the top 5 most common venues

In [148]:
num_top_venues = 5

for hood in bangalore_grouped['Neighborhood']:
    print("----",hood,"----")
    temp = bangalore_grouped[bangalore_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

---- A F station yelahanka ----
               venue  freq
0  Indian Restaurant   0.4
1   Department Store   0.2
2         Hot Spring   0.2
3               Lake   0.2
4  Accessories Store   0.0


---- Adugodi, Hosur Road ----
               venue  freq
0               Café  0.33
1                Spa  0.11
2       Dessert Shop  0.11
3  Kerala Restaurant  0.11
4        Bus Station  0.11


---- Agara, Kendriya Sadan, Koramangala, Koramangala I block, St. john's medical college ----
               venue  freq
0  Indian Restaurant  0.21
1               Café  0.16
2        Coffee Shop  0.05
3     Ice Cream Shop  0.05
4             Bakery  0.05


---- Agram, Air Force hospital ----
               venue  freq
0  Indian Restaurant  0.50
1        Pizza Place  0.25
2      Historic Site  0.25
3  Accessories Store  0.00
4  Paella Restaurant  0.00


---- Amruthahalli, Kodigehalli, Sahakaranagar P.o ----
               venue  freq
0  Indian Restaurant  0.22
1        Bus Station  0.11
2     Ice Cream 

### Let's put that into a pandas dataframe

First, let's write a function to sort the venues in descending order.


In [149]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [150]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = bangalore_grouped['Neighborhood']

for ind in np.arange(bangalore_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(bangalore_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0                               A F station yelahanka     Indian Restaurant   
1                                 Adugodi, Hosur Road                  Café   
2   Agara, Kendriya Sadan, Koramangala, Koramangal...     Indian Restaurant   
3                           Agram, Air Force hospital     Indian Restaurant   
4        Amruthahalli, Kodigehalli, Sahakaranagar P.o     Indian Restaurant   
5             Anandnagar, H.A. farm, Hebbal Kempapura     Indian Restaurant   
6   Anekal, Anekalbazar, Bestamaranahalli, Harogad...      Business Service   
7   Arabic College, Devarjeevanahalli, Nagavara, R...     Indian Restaurant   
8   Aranya Bhavan, Malleswaram, Palace Guttahalli,...     Indian Restaurant   
9   Ashoknagar, Banashankari, Dasarahalli, State B...           Snack Place   
10  Attur, Bsf Campus yelahanka, Crpf Campus yelah...                  Café   
11                            Austin Town, Viveknagar                   Spa   
12  Avalahalli, Bapujinagar, Byatarayanapura, Deep...     Indian Restaurant   
13  Avani Sringeri mutt, Basaveswaranagar Ii stage...                Bakery   
14  Avenue Road, Bangalore City, Bangalore Corpora...     Indian Restaurant   
15  B Sk ii stage, Jayanagar West, Padmanabhnagar,...                  Café   
16                             Bagalgunte, Nagasandra     Indian Restaurant   
17      Bagalur, Bandikodigehalli, Doddagubbi, Kannur           Sports Club   
18                                 Balepete, Chickpet     Indian Restaurant   
19  Banashankari Iii stage, Girinagar, Hosakerehal...        Clothing Store   
20     Banaswadi, Horamavu, Jalavayuvihar, Kalyanagar                  Café   
21                  Bangalore Air port, Nal, Vimapura     Indian Restaurant   
22  Bangalore Bazaar, Bangalore., Brigade Road, Cu...     Indian Restaurant   
23  Bangalore Dist offices bldg, K. g. road, Subha...     Indian Restaurant   
24  Bangalore Sub fgn post, Cmp Centre and school,...     Indian Restaurant   
25                                       Bannerghatta                Resort   
26  Bannerghatta Road, Hulimavu, Mico Layout, Moun...     Indian Restaurant   
27  Basavanagudi, Lalbagh West, Mavalli, Pasmpamah...     Indian Restaurant   
28     Basaveshwaranagar, K H b colony, Kamakshipalya  Fast Food Restaurant   
29                                          Bellandur                 Hotel   
30                                        Benson Town     Indian Restaurant   
31  Bettahalsur, Chikkajala, Doddajala, Hunasamara...     Indian Restaurant   
32  Bhashyam Circle, Industrial Estate, Rajajinaga...     Indian Restaurant   
33  Bhattarahalli, Bidrahalli, Medihalli, Mundur, ...                  Café   
34  Bnagalore Viswavidalaya, Mallathahalli, Ullalu...     Indian Restaurant   
35                Bommanahalli, Hongasandra, Madivala     Indian Restaurant   
36                                    C.V.raman nagar     Indian Restaurant   
37          Chamrajpet, Chamrajpet Bazar, Vittalnagar      Department Store   
38     Chandra Lay out, Vijayanagar, Vijayanagar East     Indian Restaurant   
39                   Chikkabettahalli, Vidyaranyapura     Indian Restaurant   
40                     Chikkalasandra, Subramanyapura         Movie Theater   
41  Devanagundi, Kadugodi, Kalkunte, Kannamangala,...                 Hotel   
42                       Devasandra, Krishnarajapuram                 Hotel   
43                      Dharmaram College, Taverekere     Indian Restaurant   
44                      Doddakallasandra, Konanakunte     Indian Restaurant   
45  Doddanekkundi, Kundalahalli, Marathahalli Colo...     Indian Restaurant   
46                                             Domlur     Indian Restaurant   
47        Dommasandra, Handenahalli, Kugur, Sarjapura           Snack Place   
48  Doorvaninagar, Krishnarajapuram R s, Ramamurth...     Indian Restaurant   
49                                   Electronics City                Bakery   
50          

## Cluster Neighborhoods

#### Run k-means to cluster the neighborhood into 4 clusters.


In [151]:
# set number of clusters
kclusters = 4

bangalore_grouped_clustering = bangalore_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(bangalore_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 2, 0, 0, 0, 2, 1, 0, 0, 2], dtype=int32)

In [152]:
len(kmeans.labels_)

96

In [153]:
bangalore_grouped_clustering.head()

Accessories Store  Afghan Restaurant  American Restaurant  \
0                0.0                0.0             0.000000   
1                0.0                0.0             0.000000   
2                0.0                0.0             0.017857   
3                0.0                0.0             0.000000   
4                0.0                0.0             0.000000   

   Andhra Restaurant  Arcade  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.000000     0.0          0.0         0.0                  0.0   
1           0.000000     0.0          0.0         0.0                  0.0   
2           0.017857     0.0          0.0         0.0                  0.0   
3           0.000000     0.0          0.0         0.0                  0.0   
4           0.000000     0.0          0.0         0.0                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0               0.0                 0.0          0.0            0.0   
1               0.0                 0.0          0.0            0.0   
2               0.0                 0.0          0.0            0.0   
3               0.0                 0.0          0.0            0.0   
4               0.0                 0.0          0.0            0.0   

   BBQ Joint  Badminton Court  Bagel Shop    Bakery       Bar  \
0   0.000000              0.0         0.0  0.000000  0.000000   
1   0.000000              0.0         0.0  0.000000  0.000000   
2   0.017857              0.0         0.0  0.053571  0.017857   
3   0.000000              0.0         0.0  0.000000  0.000000   
4   0.000000              0.0         0.0  0.000000  0.000000   

   Basketball Court  Bed & Breakfast  Bengali Restaurant  Bike Shop  Bistro  \
0               0.0              0.0                 0.0        0.0     0.0   
1               0.0              0.0                 0.0        0.0     0.0   
2               0.0              0.0                 0.0        0.0     0.0   
3               0.0              0.0                 0.0        0.0     0.0   
4               0.0              0.0                 0.0        0.0     0.0   

   Boarding House  Boat or Ferry  Bookstore  Botanical Garden  Boutique  \
0             0.0            0.0        0.0               0.0       0.0   
1             0.0            0.0        0.0               0.0       0.0   
2             0.0            0.0        0.0               0.0       0.0   
3             0.0            0.0        0.0               0.0       0.0   
4             0.0            0.0        0.0               0.0       0.0   

   Bowling Alley  Breakfast Spot   Brewery  Bridge  Bubble Tea Shop  \
0            0.0        0.000000  0.000000     0.0         0.000000   
1            0.0        0.000000  0.000000     0.0         0.000000   
2            0.0        0.017857  0.000000     0.0         0.000000   
3            0.0        0.000000  0.000000     0.0         0.000000   
4            0.0        0.000000  0.037037     0.0         0.037037   

   Burger Joint  Burrito Place  Bus Station  Bus Stop  Business Service  \
0      0.000000            0.0     0.000000       0.0          0.000000   
1      0.000000            0.0     0.111111       0.0          0.111111   
2      0.000000            0.0     0.000000       0.0          0.000000   
3      0.000000            0.0     0.000000       0.0          0.000000   
4      0.037037            0.0     0.111111       0.0          0.000000   

   Butcher  Cafeteria      Café  Candy Store  Cantonese Restaurant  \
0      0.0        0.0  0.000000          0.0                   0.0   
1      0.0        0.0  0.333333          0.0                   0.0   
2      0.0        0.0  0.160714          0.0                   0.0   
3      0.0        0.0  0.000000          0.0                   0.0   
4      0.0        0.0  0.037037          0.0                   0.0   

   Capitol Building  Caribbean Restaurant  Chaat Place  Chettinad Restaurant  \
0               0.0       

#### Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [154]:
bangalore_merged = bangalore_data

# add clustering labels
bangalore_merged['Cluster Labels'] = kmeans.labels_

# merge bangalore_grouped with bangalore_data to add latitude/longitude for each neighborhood
bangalore_merged = bangalore_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'),
                                                                on='Neighborhood')

bangalore_merged.head() # check the last columns!

Neighborhood  Pincode   Latitude  \
0                              A F station yelahanka   560063  13.129087   
1                                Adugodi, Hosur Road   560030  12.944150   
2  Agara, Kendriya Sadan, Koramangala, Koramangal...   560034  12.926138   
3                          Agram, Air Force hospital   560007  12.957917   
4       Amruthahalli, Kodigehalli, Sahakaranagar P.o   560092  13.064104   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  77.614226               0     Indian Restaurant            Hot Spring   
1  77.607623               2                  Café                   Spa   
2  77.622109               0     Indian Restaurant                  Café   
3  77.630912               0     Indian Restaurant           Pizza Place   
4  77.593121               0     Indian Restaurant        Ice Cream Shop   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
0      Department Store                  Lake  Eastern European Restaurant   
1          Dessert Shop           Bus Station            Kerala Restaurant   
2                Bakery        Ice Cream Shop                  Coffee Shop   
3         Historic Site   Dumpling Restaurant                  Flower Shop   
4           Bus Station                Resort                      Brewery   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0           Flower Shop           Flea Market     Fish & Chips Shop   
1            Restaurant      Business Service     Electronics Store   
2  Fast Food Restaurant                   Pub                 Hotel   
3           Flea Market     Fish & Chips Shop                 Field   
4                 Hotel    Italian Restaurant           Flea Market   

  9th Most Common Venue 10th Most Common Venue  
0                 Field   Fast Food Restaurant  
1           Flea Market      Fish & Chips Shop  
2           Pizza Place     Mexican Restaurant  
3  Fast Food Restaurant         Farmers Market  
4  Fast Food Restaurant       Department Store

In [155]:
bangalore_merged['Cluster Labels'].value_counts()

2    50
0    44
3     1
1     1
Name: Cluster Labels, dtype: int64

### Finally, let's visualize the resulting clusters

In [156]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(bangalore_merged['Latitude'], bangalore_merged['Longitude'], bangalore_merged['Neighborhood'], bangalore_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-2],
        fill_opacity=0.5).add_to(map_clusters)
       
map_clusters

## Lets Evaluate the clusters and find out potential business point for a new Restaurant

### Cluster-0

In [157]:
bangalore_merged.loc[bangalore_merged['Cluster Labels'] == 0,
                     #bangalore_merged.columns[['Pincode']],
                     bangalore_merged.columns[[0]+[1] + list(range(4, bangalore_merged.shape[1]))]]

Neighborhood  Pincode  \
0                               A F station yelahanka   560063   
2   Agara, Kendriya Sadan, Koramangala, Koramangal...   560034   
3                           Agram, Air Force hospital   560007   
4        Amruthahalli, Kodigehalli, Sahakaranagar P.o   560092   
7   Arabic College, Devarjeevanahalli, Nagavara, R...   560045   
8   Aranya Bhavan, Malleswaram, Palace Guttahalli,...   560003   
12  Avalahalli, Bapujinagar, Byatarayanapura, Deep...   560026   
14  Avenue Road, Bangalore City, Bangalore Corpora...   560002   
16                             Bagalgunte, Nagasandra   560073   
18                                 Balepete, Chickpet   560053   
21                  Bangalore Air port, Nal, Vimapura   560017   
23  Bangalore Dist offices bldg, K. g. road, Subha...   560009   
26  Bannerghatta Road, Hulimavu, Mico Layout, Moun...   560076   
27  Basavanagudi, Lalbagh West, Mavalli, Pasmpamah...   560004   
31  Bettahalsur, Chikkajala, Doddajala, Hunasamara...   562157   
32  Bhashyam Circle, Industrial Estate, Rajajinaga...   560010   
34  Bnagalore Viswavidalaya, Mallathahalli, Ullalu...   560056   
35                Bommanahalli, Hongasandra, Madivala   560068   
36                                    C.V.raman nagar   560093   
39                   Chikkabettahalli, Vidyaranyapura   560097   
44                      Doddakallasandra, Konanakunte   560062   
45  Doddanekkundi, Kundalahalli, Marathahalli Colo...   560037   
46                                             Domlur   560071   
48  Doorvaninagar, Krishnarajapuram R s, Ramamurth...   560016   
51                                           G.K.v.k.   560065   
52  Gaviopuram Extension, Gaviopuram Guttanahalli,...   560019   
55  Govindapalya, H M t, Jalahalli, Jalahalli Village   560013   
58                                        H.K.p. road   560051   
64  Isro Anthariksha bhavan, R.M.v. extension ii s...   560094   
65  J P nagar, Jp Nagar iii phase, Kumaraswamy Lay...   560078   
67                                     Jalahalli East   560014   
69                              Jayanagar, Tilaknagar   560041   
70                                      Jayangar East   560069   
71                  Jayangar Iii block, Madhavan Park   560011   
73  Kacharakanahalli, Lingarajapuram, St. thomas town   560084   
79                       M S r road, Mathikere, Msrit   560054   
80                                        Magadi Road   560023   
83                                        Muthanallur   560099   
86                                   New Thippasandra   560075   
88                                         Rv Niketan   560059   
90      Sampangiramnagar, Shanthinagar, Wilson Garden   560027   
92                               Sivan Chetty gardens   560042   
93                                       Tyagrajnagar   560028   
94                                       Vasanthnagar   560052   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0                0     Indian Restaurant            Hot Spring   
2                0     Indian Restaurant                  Café   
3                0     Indian Restaurant           Pizza Place   
4                0     Indian Restaurant        Ice Cream Shop   
7                0     Indian Restaurant           Coffee Shop   
8                0     Indian Restaurant        Ice Cream Shop   
12               0     Indian Restaurant           Pizza Place   
14               0     Indian Restaurant               Theater   
16               0     Indian Restaurant    Light Rail Station   
18               0     Indian Restaurant             Bookstore   
21               0     Indian Restaurant  Fast Food Restaurant   
23               0     Indian Restaurant                 Hotel   
26               0     Indian Restaurant   Sporting Goods Shop   
27               0     Indian Restaurant  Fast Food Restaurant   
31               0     Indian Restaurant                 Hotel   
32            

### Cluster-1

In [158]:
bangalore_merged.loc[bangalore_merged['Cluster Labels'] == 1, 
                     bangalore_merged.columns[[0]+[1] + list(range(5, bangalore_merged.shape[1]))]]

Neighborhood  Pincode  \
6  Anekal, Anekalbazar, Bestamaranahalli, Harogad...   562106   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
6      Business Service           Yoga Studio     Convenience Store   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
6                  Food           Flower Shop           Flea Market   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
6     Fish & Chips Shop                 Field  Fast Food Restaurant   

  10th Most Common Venue  
6         Farmers Market

### Cluster-2

In [159]:
bangalore_merged.loc[bangalore_merged['Cluster Labels'] == 2, 
                     bangalore_merged.columns[[0]+[1] + list(range(5, bangalore_merged.shape[1]))]]

Neighborhood  Pincode  \
1                                 Adugodi, Hosur Road   560030   
5             Anandnagar, H.A. farm, Hebbal Kempapura   560024   
9   Ashoknagar, Banashankari, Dasarahalli, State B...   560050   
10  Attur, Bsf Campus yelahanka, Crpf Campus yelah...   560064   
11                            Austin Town, Viveknagar   560047   
13  Avani Sringeri mutt, Basaveswaranagar Ii stage...   560086   
15  B Sk ii stage, Jayanagar West, Padmanabhnagar,...   560070   
17      Bagalur, Bandikodigehalli, Doddagubbi, Kannur   562149   
19  Banashankari Iii stage, Girinagar, Hosakerehal...   560085   
20     Banaswadi, Horamavu, Jalavayuvihar, Kalyanagar   560043   
22  Bangalore Bazaar, Bangalore., Brigade Road, Cu...   560001   
24  Bangalore Sub fgn post, Cmp Centre and school,...   560025   
28     Basaveshwaranagar, K H b colony, Kamakshipalya   560079   
29                                          Bellandur   560103   
30                                        Benson Town   560046   
33  Bhattarahalli, Bidrahalli, Medihalli, Mundur, ...   560049   
37          Chamrajpet, Chamrajpet Bazar, Vittalnagar   560018   
38     Chandra Lay out, Vijayanagar, Vijayanagar East   560040   
40                     Chikkalasandra, Subramanyapura   560061   
41  Devanagundi, Kadugodi, Kalkunte, Kannamangala,...   560067   
42                       Devasandra, Krishnarajapuram   560036   
43                      Dharmaram College, Taverekere   560029   
47        Dommasandra, Handenahalli, Kugur, Sarjapura   562125   
49                                   Electronics City   560100   
50                          Fraser Town, Jeevanahalli   560005   
53  Gayathrinagar, Okalipuram, Ramachandrapuram, S...   560021   
54   Goraguntepalya, Yeshwanthpur Bazar, Yeswanthpura   560022   
56              Gunjur, Muthusandra, Panathur, Vartur   560087   
57       H.A.l ii stage, Hulsur Bazaar, Someswarapura   560008   
59                                         Hampinagar   560104   
60                                Hoodi, Mahadevapura   560048   
61                                         Hsr Layout   560102   
62  Immedihalli, Jeevanbhimanagar, Ramagondanahall...   560066   
63              Indiranagar, Indiranagar Com. complex   560038   
66                    J.C.nagar, Training Command iaf   560006   
68                   Jalahalli West, Kamagondanahalli   560015   
72                            K.P.west, Seshadripuram   560020   
74                     Kanteeravanagar, Nandinilayout   560096   
75                 Kenchanahalli, Rajarajeshwarinagar   560098   
76                                  Koramangala Vi bk   560095   
77                                           Kothanur   560077   
78                                         Kumbalgodu   560074   
81   Malkand Lines, Mandalay Lines, Maruthi Sevanagar   560033   
82                      Malleswaram West, Milk Colony   560055   
84                                         Nagarbhavi   560072   
85                                       Nayandahalli   560039   
87                P&t Col. kavalbyrasandra, R T nagar   560032   
89                                     Sadashivanagar   560080   
91                                  Science Institute   560012   
95                                        Viswaneedam   560091   

   1st Most Common Venue   2nd Most Common Venue  \
1                   Café                     Spa   
5      Indian Restaurant                  Bakery   
9            Snack Place       Indian Restaurant   
10                  Café       Indian Restaurant   
11                   Spa                    Park   
13                Bakery             Pizza Place   
15                  Café       Indian Restaurant   
17           Sports Club             Yoga Studio   
19        Clothing Store               Juice Bar   
20                  Café       Indian Restaurant   
22     Indian Restaurant                   Hotel   
24     Indian Restaurant                    Café  

### Cluster-3

In [160]:
bangalore_merged.loc[bangalore_merged['Cluster Labels'] == 3, 
                     bangalore_merged.columns[[0]+[1] + list(range(5, bangalore_merged.shape[1]))]]

Neighborhood  Pincode 1st Most Common Venue 2nd Most Common Venue  \
25  Bannerghatta   560083                Resort           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
25   Dumpling Restaurant           Flower Shop           Flea Market   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
25     Fish & Chips Shop                 Field  Fast Food Restaurant   

   9th Most Common Venue 10th Most Common Venue  
25        Farmers Market            Event Space

## Observation & Evaluation ->

### Cluster 1 & 3 -                
            There are only one 'Pincode Area' in each of these 2 clusters and they are
            looking perfect for opening a restaurant as there are very less no. of 
            restaurants and other places like business areas and flea markets which
            are indication of potential customer based areas for restaurant.
            But this is good if someone is looking to open a restaurants in outskirts
            of bangalore as these locations are very far from the actual city market area.
### Cluster 0 -
            These neighborhoods have lot of restaurants and eateries. Although these 
            places would have high business turn around thats why there are so many 
            eateries already established, but competition will be very very tough 
            for a new restaurant to open and settle.
### Cluster 2 -
            These neighborhoods have mix of all types of venues, lets explore and analyse
            which areas have less restaurants and eateries and other business venues which
            will fetch potential customers for a restaurant.

### Lets create the Cluster2 dataframce and analyze further

In [161]:
cl2 = bangalore_merged.loc[bangalore_merged['Cluster Labels'] == 2, 
                     bangalore_merged.columns[[0]+[1]+[2]+[3] + list(range(5,bangalore_merged.shape[1]))]]

cl2.reset_index(drop=True, inplace = True)

print(cl2.shape)
cl2.head()

(50, 14)


Neighborhood  Pincode   Latitude  \
0                                Adugodi, Hosur Road   560030  12.944150   
1            Anandnagar, H.A. farm, Hebbal Kempapura   560024  13.048745   
2  Ashoknagar, Banashankari, Dasarahalli, State B...   560050  12.935307   
3  Attur, Bsf Campus yelahanka, Crpf Campus yelah...   560064  13.101970   
4                            Austin Town, Viveknagar   560047  12.953679   

   Longitude 1st Most Common Venue 2nd Most Common Venue  \
0  77.607623                  Café                   Spa   
1  77.592330     Indian Restaurant                Bakery   
2  77.558531           Snack Place     Indian Restaurant   
3  77.586359                  Café     Indian Restaurant   
4  77.617625                   Spa                  Park   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0          Dessert Shop           Bus Station     Kerala Restaurant   
1           Pizza Place          Burger Joint                  Lake   
2      Department Store  Fast Food Restaurant        Breakfast Spot   
3                Bakery    Chinese Restaurant      Department Store   
4  Fast Food Restaurant                  Café   Indie Movie Theater   

  6th Most Common Venue          7th Most Common Venue  \
0            Restaurant               Business Service   
1      Department Store                       Pharmacy   
2                  Park  Vegetarian / Vegan Restaurant   
3            Restaurant                     Smoke Shop   
4        Cricket Ground    Eastern European Restaurant   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Electronics Store           Flea Market      Fish & Chips Shop  
1                     Café           Bus Station         Ice Cream Shop  
2                Juice Bar                Arcade                 Bakery  
3  South Indian Restaurant            Food Truck           Soccer Field  
4              Flower Shop           Flea Market      Fish & Chips Shop

#### List out neighborhoods don't have "Indian Restaurant" in their 10 most common places
As we have seen in the cluster-1 * as well as in the cluster-2, the venue category
"Indian Restaurant" is most popular and common places in most of the neighborhoods. So Let's find out the neighborhoods where there are no "Indian Restaurant" venues in the 10 most common venues.

In [162]:
cl2noIR = cl2[cl2.apply(lambda r: r.str.contains('Indian Restaurant', case=False).any(), axis=1) == False]

print('We found {} neighboorhood areas not having Indian Restaurant as their 10 most common venues'.
      format(cl2noIR.shape[0]))

cl2noIR.reset_index(drop=True, inplace=True)
cl2noIR.head()

We found 18 neighboorhood areas not having Indian Restaurant as their 10 most common venues


Neighborhood  Pincode   Latitude  \
0                                Adugodi, Hosur Road   560030  12.944150   
1                            Austin Town, Viveknagar   560047  12.953679   
2      Bagalur, Bandikodigehalli, Doddagubbi, Kannur   562149  13.119511   
3                     Chikkalasandra, Subramanyapura   560061  12.903725   
4  Devanagundi, Kadugodi, Kalkunte, Kannamangala,...   560067  12.996701   

   Longitude 1st Most Common Venue 2nd Most Common Venue  \
0  77.607623                  Café                   Spa   
1  77.617625                   Spa                  Park   
2  77.678614           Sports Club           Yoga Studio   
3  77.537639         Movie Theater                  Café   
4  77.758197                 Hotel           Pizza Place   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0          Dessert Shop           Bus Station     Kerala Restaurant   
1  Fast Food Restaurant                  Café   Indie Movie Theater   
2     Food & Drink Shop           Flower Shop           Flea Market   
3            Restaurant   Dumpling Restaurant           Flower Shop   
4         Train Station           Bus Station           Yoga Studio   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
0            Restaurant             Business Service     Electronics Store   
1        Cricket Ground  Eastern European Restaurant           Flower Shop   
2     Fish & Chips Shop                        Field  Fast Food Restaurant   
3           Flea Market            Fish & Chips Shop                 Field   
4   Dumpling Restaurant                  Flea Market     Fish & Chips Shop   

  9th Most Common Venue 10th Most Common Venue  
0           Flea Market      Fish & Chips Shop  
1           Flea Market      Fish & Chips Shop  
2        Farmers Market            Event Space  
3  Fast Food Restaurant         Farmers Market  
4                 Field   Fast Food Restaurant

#### Now let's get the density of the venue category "Restaurant" for each neighborhood
    For this we will check common venue categories for each neighborhood, if its any 
    kind of Restaurant then multiply it with its corresponding weightage value 
    (1st common venue = 10 & 10th common venue = 1)then sum up all the weightage values.

In [163]:
rDensityList = []
for index, row in cl2noIR.iterrows():
    counter = 10
    restaurantDensity = 0
    for col in cl2noIR.iloc[:,4:15]:
#        print(counter, cl2noIR.loc[index,col])
        venue_type = cl2noIR.loc[index,col]
        if 'Restaurant' in venue_type:
            restaurantDensity += counter
        counter -= 1
#    print('\n', "Restaurant Weight ->", restaurantDensity, '\n')
    rDensityList.append(restaurantDensity)
rDensityList

[11, 12, 3, 17, 6, 5, 9, 0, 9, 11, 0, 0, 5, 9, 12, 7, 0, 15]

#### Add the Restaurant Density column into the dataframe

In [164]:
cl2noIR['Restaurant Density'] = rDensityList
#cl2noIR.loc[:,'Restaurant Density'] = rDensityList # This is another method
cl2noIR.shape

/home/jupyterlab/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


(18, 15)

#### Sort the Neighborhoods according to the Restaurant Density

In [165]:
cl2noIR.loc[:,('Neighborhood','Restaurant Density')].sort_values(['Restaurant Density'])

Neighborhood  Restaurant Density
7               Gunjur, Muthusandra, Panathur, Vartur                   0
16                                       Nayandahalli                   0
10                 Kenchanahalli, Rajarajeshwarinagar                   0
11                                  Koramangala Vi bk                   0
2       Bagalur, Bandikodigehalli, Doddagubbi, Kannur                   3
5                        Devasandra, Krishnarajapuram                   5
12                                           Kothanur                   5
4   Devanagundi, Kadugodi, Kalkunte, Kannamangala,...                   6
15                                         Nagarbhavi                   7
8                                          Hampinagar                   9
13                                         Kumbalgodu                   9
6         Dommasandra, Handenahalli, Kugur, Sarjapura                   9
0                                 Adugodi, Hosur Road                  11
9                    Jalahalli West, Kamagondanahalli                  11
14   Malkand Lines, Mandalay Lines, Maruthi Sevanagar                  12
1                             Austin Town, Viveknagar                  12
17                                        Viswaneedam                  15
3                      Chikkalasandra, Subramanyapura                  17

In [166]:
cl2noIR.loc[0,'Restaurant Density']

11

#### Lets create the final dataframe for the recommended neighborhoods with Restaurant Density less than 5 to set up the new Restaurant.

In [176]:
bFinalDF = cl2noIR[cl2noIR[('Restaurant Density')] < 3].sort_values(['Restaurant Density'])
bFinalDF.reset_index(drop=True, inplace=True)

bFinalDF[['Neighborhood','Restaurant Density']]

Neighborhood  Restaurant Density
0  Gunjur, Muthusandra, Panathur, Vartur                   0
1     Kenchanahalli, Rajarajeshwarinagar                   0
2                      Koramangala Vi bk                   0
3                           Nayandahalli                   0

### Plot the Map to show the recomended neighborhoods in the Bangalore Map 

In [178]:
# create map of Bangalore using latitude and longitude values

map_recomend = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, neighborhood, pincode in zip(bFinalDF['Latitude'], bFinalDF['Longitude'],
                                  bFinalDF['Neighborhood'], bFinalDF['Pincode']):
    label = '{}'', ''{}'.format(neighborhood,pincode)
#    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.5,
#        parse_html=False).add_to(map_recomend)
        ).add_to(map_recomend) 
    
map_recomend

## Final Recomendation of Neighborhoods For A Restaurant & Type of Restaurant
    As per the data available from mapofindia.com website & foursquare API,
    After the plotting the final set of neighborhoods in the Map above, we can confidently say 
    that below two in our list are spot on & are near around the core city as well for opening a
    restaurant
###    1. Koramangala Vi bk 	 - 560095
###    2. Nayandahalli - 560039

### Restaurant Type - Indian Restaurant

     
     
   
### ^^^^^^^^^^^^^^^^    Thank You Very Much Reviewing This    ^^^^^^^^^^^^^^^^